File: examples/partner-api/proactive/python/proactive_messaging.py
#!/usr/bin/env python3

"""
Nexo Partner API - Proactive Messaging Example (Python)

This example demonstrates how to send proactive messages to subscribers
using the Nexo Partner API. It covers:
1. Listing subscribers
2. Getting subscriber threads
3. Sending a proactive message

Use case: Delivery service arrival notification

Note: Proactive messaging uses the same auth headers (X-App-Id, X-App-Secret)
but the POST /apps/{app_id}/threads/{thread_id}/messages endpoint shape is
separate from the webhook contract.
"""

import os
import sys
from pathlib import Path
from typing import Dict, List, Optional, Any

import requests
from dotenv import load_dotenv

# Load environment variables from parent directory
env_path = Path(__file__).parent.parent / ".env"
load_dotenv(dotenv_path=env_path)


# Configuration from environment
class Config:
    """API configuration loaded from environment variables"""

    base_url: str = os.getenv("API_BASE_URL", "http://localhost:8000")
    app_id: Optional[str] = os.getenv("APP_ID")
    app_secret: Optional[str] = os.getenv("APP_SECRET")
    subscriber_id: Optional[str] = os.getenv("SUBSCRIBER_ID")
    thread_id: Optional[str] = os.getenv("THREAD_ID")

    @classmethod
    def validate(cls) -> None:
        """Validate required configuration"""
        if not cls.app_id or not cls.app_secret:
            print("❌ Error: APP_ID and APP_SECRET are required.")
            print("Please copy .env.example to .env and fill in your credentials.")
            sys.exit(1)


class PartnerAPIClient:
    """Client for Nexo Partner API with authentication"""

    def __init__(self, base_url: str, app_id: str, app_secret: str):
        self.base_url = base_url.rstrip("/")
        self.app_id = app_id
        self.app_secret = app_secret
        self.session = requests.Session()
        self.session.headers.update(
            {
                "X-App-Id": app_id,
                "X-App-Secret": app_secret,
                "Content-Type": "application/json",
            }
        )

    def _make_request(
        self, method: str, endpoint: str, data: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Make an authenticated API request

        Args:
            method: HTTP method (GET, POST, etc.)
            endpoint: API endpoint path
            data: Request body data (for POST requests)

        Returns:
            API response data

        Raises:
            Exception: If API request fails
        """
        url = f"{self.base_url}{endpoint}"

        try:
            response = self.session.request(method, url, json=data)
            response.raise_for_status()
            return response.json()

        except requests.exceptions.HTTPError as e:
            status = e.response.status_code
            try:
                message = e.response.json().get("error", e.response.text)
            except Exception:
                message = e.response.text

            if status == 401:
                raise Exception(
                    f"Authentication failed: {message}. "
                    "Check your APP_ID and APP_SECRET."
                )
            elif status == 404:
                raise Exception(f"Resource not found: {message}")
            elif status == 500:
                raise Exception(f"Server error: {message}. Please try again later.")
            else:
                raise Exception(f"API error ({status}): {message}")

        except requests.exceptions.ConnectionError:
            raise Exception(
                f"Network error: Could not reach {self.base_url}. "
                "Is the server running?"
            )
        except requests.exceptions.RequestException as e:
            raise Exception(f"Request error: {str(e)}")

    def list_subscribers(self) -> List[Dict[str, Any]]:
        """
        List all subscribers for this app

        Returns:
            List of subscriber objects
        """
        print("📋 Fetching subscribers...")
        data = self._make_request("GET", f"/api/apps/{self.app_id}/subscribers")
        return data.get("subscribers", [])

    def get_subscriber_threads(self, subscriber_id: str) -> List[Dict[str, Any]]:
        """
        Get all conversation threads for a subscriber

        Args:
            subscriber_id: The subscriber's ID

        Returns:
            List of thread objects
        """
        print(f"💬 Fetching threads for subscriber {subscriber_id}...")
        data = self._make_request(
            "GET", f"/api/apps/{self.app_id}/subscribers/{subscriber_id}/threads"
        )
        return data.get("threads", [])

    def send_message(self, thread_id: str, message: Dict[str, str]) -> Dict[str, Any]:
        """
        Send a proactive message to a thread

        Args:
            thread_id: The thread ID to send the message to
            message: Message object with role and content

        Returns:
            Created message object
        """
        print(f"📤 Sending message to thread {thread_id}...")
        return self._make_request(
            "POST",
            f"/api/apps/{self.app_id}/threads/{thread_id}/messages",
            data=message,
        )


def main() -> None:
    """Main example: Send a delivery arrival notification"""
    print("🚀 Nexo Partner API - Proactive Messaging Example\n")

    # Validate configuration
    Config.validate()

    # Initialize API client
    client = PartnerAPIClient(Config.base_url, Config.app_id, Config.app_secret)

    try:
        # Step 1: List subscribers (or use provided SUBSCRIBER_ID)
        target_subscriber_id = Config.subscriber_id

        if not target_subscriber_id:
            subscribers = client.list_subscribers()
            print(f"✅ Found {len(subscribers)} subscriber(s)\n")

            if not subscribers:
                print(
                    "⚠️  No subscribers found. Users need to authorize your app first."
                )
                return

            # Use the first subscriber for this example
            target_subscriber_id = subscribers[0]["id"]
            print(f"📌 Using subscriber: {target_subscriber_id}\n")

        # Step 2: Get subscriber's threads (or use provided THREAD_ID)
        target_thread_id = Config.thread_id

        if not target_thread_id:
            threads = client.get_subscriber_threads(target_subscriber_id)
            print(f"✅ Found {len(threads)} thread(s)\n")

            if not threads:
                print(
                    "⚠️  No active threads found. The subscriber needs to have an existing conversation."
                )
                return

            # Use the first thread for this example
            target_thread_id = threads[0]["id"]
            print(f"📌 Using thread: {target_thread_id}\n")

        # Step 3: Send a proactive message
        # Use case: Delivery service arrival notification
        delivery_id = "DLV-78910"
        driver_name = "Alex"

        message = {
            "role": "assistant",
            "content": (
                f"🚚 Your delivery is arriving soon!\n\n"
                f"Delivery ID: {delivery_id}\n"
                f"Driver: {driver_name}\n"
                f"ETA: 15 minutes\n\n"
                f"Please be available to receive your package. "
                f"You can track your driver in real-time through the app.\n\n"
                f"Need to give special delivery instructions? Just reply to this message!"
            ),
        }

        sent_message = client.send_message(target_thread_id, message)
        print("✅ Message sent successfully!\n")
        print("Message details:")
        print(f"  ID: {sent_message['id']}")
        print(f"  Thread: {sent_message['thread_id']}")
        print(f"  Content: {sent_message['content'][:50]}...")
        print(f"  Created: {sent_message['created_at']}\n")

        print("🎉 Example completed successfully!")

    except Exception as e:
        print(f"\n❌ Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: examples/partner-api/proactive/typescript/proactive-messaging.js
#!/usr/bin/env node

/**
 * Nexo Partner API - Proactive Messaging Example (Node.js)
 *
 * This example demonstrates how to send proactive messages to subscribers
 * using the Nexo Partner API. It covers:
 * 1. Listing subscribers
 * 2. Getting subscriber threads
 * 3. Sending a proactive message
 *
 * Use case: E-commerce order shipping notification
 *
 * Note: Proactive messaging uses the same auth headers (X-App-Id, X-App-Secret)
 * but the POST /apps/{app_id}/threads/{thread_id}/messages endpoint shape is
 * separate from the webhook contract.
 */

import axios from 'axios';
import dotenv from 'dotenv';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

// Load environment variables from parent directory
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
dotenv.config({ path: join(__dirname, '..', '.env') });

// Configuration from environment
const config = {
  baseUrl: process.env.API_BASE_URL || 'http://localhost:8000',
  appId: process.env.APP_ID,
  appSecret: process.env.APP_SECRET,
  subscriberId: process.env.SUBSCRIBER_ID,
  threadId: process.env.THREAD_ID,
};

// Validate required configuration
if (!config.appId || !config.appSecret) {
  console.error('❌ Error: APP_ID and APP_SECRET are required.');
  console.error('Please copy .env.example to .env and fill in your credentials.');
  process.exit(1);
}

/**
 * Make an authenticated API request to the Partner API
 * @param {string} method - HTTP method (GET, POST, etc.)
 * @param {string} endpoint - API endpoint path
 * @param {object} data - Request body data (for POST requests)
 * @returns {Promise<object>} API response data
 */
async function makeApiRequest(method, endpoint, data = null) {
  try {
    const response = await axios({
      method,
      url: `${config.baseUrl}${endpoint}`,
      headers: {
        'X-App-Id': config.appId,
        'X-App-Secret': config.appSecret,
        'Content-Type': 'application/json',
      },
      data,
    });

    return response.data;
  } catch (error) {
    if (error.response) {
      // Server responded with error status
      const status = error.response.status;
      const message = error.response.data?.error || error.response.statusText;

      switch (status) {
        case 401:
          throw new Error(`Authentication failed: ${message}. Check your APP_ID and APP_SECRET.`);
        case 404:
          throw new Error(`Resource not found: ${message}`);
        case 500:
          throw new Error(`Server error: ${message}. Please try again later.`);
        default:
          throw new Error(`API error (${status}): ${message}`);
      }
    } else if (error.request) {
      // Request made but no response received
      throw new Error(`Network error: Could not reach ${config.baseUrl}. Is the server running?`);
    } else {
      // Something else happened
      throw new Error(`Request error: ${error.message}`);
    }
  }
}

/**
 * List all subscribers for this app
 * @returns {Promise<Array>} List of subscriber objects
 */
async function listSubscribers() {
  console.log('📋 Fetching subscribers...');
  const data = await makeApiRequest('GET', `/apps/${config.appId}/subscribers`);
  return data.subscribers || [];
}

/**
 * Get all conversation threads for a subscriber
 * @param {string} subscriberId - The subscriber's ID
 * @returns {Promise<Array>} List of thread objects
 */
async function getSubscriberThreads(subscriberId) {
  console.log(`💬 Fetching threads for subscriber ${subscriberId}...`);
  const data = await makeApiRequest('GET', `/apps/${config.appId}/subscribers/${subscriberId}/threads`);
  return data.threads || [];
}

/**
 * Send a proactive message to a thread
 * @param {string} threadId - The thread ID to send the message to
 * @param {object} message - Message object with role and content
 * @returns {Promise<object>} Created message object
 */
async function sendMessage(threadId, message) {
  console.log(`📤 Sending message to thread ${threadId}...`);
  return await makeApiRequest('POST', `/apps/${config.appId}/threads/${threadId}/messages`, message);
}

/**
 * Main example: Send an order shipping notification
 */
async function main() {
  console.log('🚀 Nexo Partner API - Proactive Messaging Example\n');

  try {
    // Step 1: List subscribers (or use provided SUBSCRIBER_ID)
    let targetSubscriberId = config.subscriberId;

    if (!targetSubscriberId) {
      const subscribers = await listSubscribers();
      console.log(`✅ Found ${subscribers.length} subscriber(s)\n`);

      if (subscribers.length === 0) {
        console.log('⚠️  No subscribers found. Users need to authorize your app first.');
        return;
      }

      // Use the first subscriber for this example
      targetSubscriberId = subscribers[0].id;
      console.log(`📌 Using subscriber: ${targetSubscriberId}\n`);
    }

    // Step 2: Get subscriber's threads (or use provided THREAD_ID)
    let targetThreadId = config.threadId;

    if (!targetThreadId) {
      const threads = await getSubscriberThreads(targetSubscriberId);
      console.log(`✅ Found ${threads.length} thread(s)\n`);

      if (threads.length === 0) {
        console.log('⚠️  No active threads found. The subscriber needs to have an existing conversation.');
        return;
      }

      // Use the first thread for this example
      targetThreadId = threads[0].id;
      console.log(`📌 Using thread: ${targetThreadId}\n`);
    }

    // Step 3: Send a proactive message
    // Use case: E-commerce order shipping notification
    const orderNumber = '12345';
    const trackingUrl = 'https://tracking.example.com/12345';

    const message = {
      role: 'assistant',
      content: `📦 Great news! Your order #${orderNumber} has shipped and is on its way!\n\n` +
               `Track your package: ${trackingUrl}\n\n` +
               `Estimated delivery: 2-3 business days.\n\n` +
               `Questions? Just ask me anytime!`,
    };

    const sentMessage = await sendMessage(targetThreadId, message);
    console.log('✅ Message sent successfully!\n');
    console.log('Message details:');
    console.log(`  ID: ${sentMessage.id}`);
    console.log(`  Thread: ${sentMessage.thread_id}`);
    console.log(`  Content: ${sentMessage.content.substring(0, 50)}...`);
    console.log(`  Created: ${sentMessage.created_at}\n`);

    console.log('🎉 Example completed successfully!');

  } catch (error) {
    console.error('\n❌ Error:', error.message);
    process.exit(1);
  }
}

// Run the example
main();
